# Project: Wrangle and Analyze WeRateDogs twitter data

In [86]:
import pandas as pd
import numpy as np
import requests
import re
import os
import tweepy
import json
import matplotlib.pyplot as plt
from scipy import stats

## Gather Data

In [87]:
# Read the csv file
df_twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    
3  891689557279858688 NaN                    NaN                    
4  891327558926688256 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0  13                10                  Phineas   None  None    None   None  
1  13                10                  Tilly     None  None    None   None  
2  12                10                  Archie    None  None    None   None  
3  13                10                  Darla     None  None    None   None  
4  12                10                  Franklin  None  None    None   None

In [88]:
# download the tsv file using the Requests library and the given URL
url =' https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)
with open('image-predictions.tsv', 'wb') as f:
        f.write(r.content)
df_image_predictions = pd.read_csv('image-predictions.tsv', sep = '\t')
df_image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh_springer_spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature_pinscher   
2  1        German_shepherd         0.596461  True    malinois             
3  1        Rhodesian_ridgeback     0.408143  True    redbone              
4  1        miniature_pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland_sheepdog    0.061428  True    
1  0.074192  True    Rhodesian_ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature_pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

##### Fetching Data from the Twitter API
Since it is a long process I put this code into a separate python file called: `fetch_twitter_data.py`

In [89]:

# # query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file
# consumer_key = 'X'
# consumer_secret = 'X'
# access_token = 'X'
# access_secret = 'X'

# auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
# auth.set_access_token(access_token, access_secret)
# api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)


After getting the data from twitter API write tweets into a file called `tweet_json.txt`

In [90]:
# with open('tweet_json.txt', 'a', encoding='utf8') as f:
#     for tweet_id in df_twitter_archive['tweet_id']:
#         try:
#             tweet = api.get_status(tweet_id)
#             json.dump(tweet._json, f)
#             f.write('\n')
#         except:
#             print("Error on tweet: " + str(tweet_id))
#             continue


In [91]:
# Create a dataframe with tweet ID, retweet count, and favorite count
df_list = []
with open('tweet_json.txt') as file:
    for line in file:
        tweet = json.loads(line)
        
        df_list.append({
            'tweet_id': tweet['id'],
            'retweet_count': tweet['retweet_count'],
            'favorite_count': tweet['favorite_count']
        })
df_tweet = pd.DataFrame(df_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])
       
df_tweet.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193  8315           38014         
1  892177421306343426  6142           32641         
2  891815181378084864  4067           24565         
3  891689557279858688  8450           41365         
4  891327558926688256  9157           39551

## Assess Data

In [92]:
df_twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
19    888202515573088257 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
...                  ...  ..                     ..                    
2326  666411507551481857 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2328  666396247373291520 NaN                    NaN                    
2329  666373753744588802 NaN                    NaN                    
2330  666362758909284353 NaN                    NaN                    
2331  666353288456101888 NaN                    NaN                    
2332  666345417576210432 NaN                    NaN                    
2333  666337882303524864 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    
2336  666273097616637952 NaN                    NaN                    
2337  666268910803644416 NaN                    NaN                    
2338  666104133288665088 NaN                    NaN                    
2339  666102155909144576 NaN                    NaN                    
2340  666099513787052032 NaN                    NaN                    
2341  666094000022159362 NaN                    NaN                    
2342  666082916733198337 NaN                    NaN                    
2343  666073100786774016 NaN                    NaN                    
2344  666071193221509120 NaN                    NaN                    
2345  666063827256086533 NaN                    NaN                    
2346  666058600524156928 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN                    
2348  666055525042405380 NaN                    NaN                    
2349  666051853826850816 NaN                    NaN     

In [93]:
df_twitter_archive[df_twitter_archive.in_reply_to_status_id.notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600  8.862664e+17           2.281182e+09          
55    881633300179243008  8.816070e+17           4.738443e+07          
64    879674319642796034  8.795538e+17           3.105441e+09          
113   870726314365509632  8.707262e+17           1.648776e+07          
148   863427515083354112  8.634256e+17           7.759620e+07          
149   863079547188785154  6.671522e+17           4.196984e+09          
179   857214891891077121  8.571567e+17           1.806710e+08          
184   856526610513747968  8.558181e+17           4.196984e+09          
186   856288084350160898  8.562860e+17           2.792810e+08          
188   855862651834028034  8.558616e+17           1.943518e+08          
189   855860136149123072  8.558585e+17           1.361572e+07          
218   850333567704068097  8.503288e+17           2.195506e+07          
228   848213670039564288  8.482121e+17           4.196984e+09          
234   847617282490613760  8.476062e+17           4.196984e+09          
251   844979544864018432  7.590995e+17           4.196984e+09          
274   840698636975636481  8.406983e+17           8.405479e+17          
290   838150277551247360  8.381455e+17           2.195506e+07          
291   838085839343206401  8.380855e+17           2.894131e+09          
313   835246439529840640  8.352460e+17           2.625958e+07          
342   832088576586297345  8.320875e+17           3.058208e+07          
346   831926988323639298  8.319030e+17           2.068372e+07          
387   826598799820865537  8.265984e+17           4.196984e+09          
409   823333489516937216  8.233264e+17           1.582854e+09          
427   821153421864615936  8.211526e+17           1.132119e+08          
498   813130366689148928  8.131273e+17           4.196984e+09          
513   811647686436880384  8.116272e+17           4.196984e+09          
565   802265048156610565  7.331095e+17           4.196984e+09          
570   801854953262350336  8.018543e+17           1.185634e+07          
576   800859414831898624  8.008580e+17           2.918590e+08          
611   797165961484890113  7.971238e+17           2.916630e+07          
...                  ...           ...                    ...          
1479  693582294167244802  6.935722e+17           1.198989e+09          
1497  692423280028966913  6.924173e+17           4.196984e+09          
1501  692142790915014657  6.920419e+17           4.196984e+09          
1523  690607260360429569  6.903413e+17           4.670367e+08          
1598  686035780142297088  6.860340e+17           4.196984e+09          
1605  685681090388975616  6.855479e+17           4.196984e+09          
1618  684969860808454144  6.849598e+17           4.196984e+09          
1630  684538444857667585  6.844811e+17           4.196984e+09          
1634  684225744407494656  6.842229e+17           4.196984e+09          
1663  682808988178739200  6.827884e+17           4.196984e+09          
1689  681340665377193984  6.813394e+17           4.196984e+09          
1774  678023323247357953  6.780211e+17           4.196984e+09          
1819  676590572941893632  6.765883e+17           4.196984e+09          
1842  675870721063669760  6.757073e+17           4.196984e+09          
1844  675849018447167488  6.758457e+17           4.196984e+09          
1852  675707330206547968  6.754971e+17           4.196984e+09          
1866  675349384339542016  6.749998e+17           4.196984e+09          
1882  674999807681908736  6.747934e+17           4.196984e+09          
1885  674793399141146624  6.717299e+17           4.196984e+09          
1892  674754018082705410  6.747522e+17           4.196984e+09          
1895  674742531037511680  6.747400e+17           4.196984e+09          
1905  674606911342424069  6.744689e+17           4.196984e+09          
1914  674330906434379776  6.658147e+17           1.637468e+07          
1940  673716320723169284  6.737159e+17           4.19698

In [94]:
df_twitter_archive[df_twitter_archive.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600  8.862664e+17           2.281182e+09          
55    881633300179243008  8.816070e+17           4.738443e+07          
64    879674319642796034  8.795538e+17           3.105441e+09          
113   870726314365509632  8.707262e+17           1.648776e+07          
148   863427515083354112  8.634256e+17           7.759620e+07          
179   857214891891077121  8.571567e+17           1.806710e+08          
185   856330835276025856 NaN                    NaN                    
186   856288084350160898  8.562860e+17           2.792810e+08          
188   855862651834028034  8.558616e+17           1.943518e+08          
189   855860136149123072  8.558585e+17           1.361572e+07          
218   850333567704068097  8.503288e+17           2.195506e+07          
228   848213670039564288  8.482121e+17           4.196984e+09          
234   847617282490613760  8.476062e+17           4.196984e+09          
274   840698636975636481  8.406983e+17           8.405479e+17          
290   838150277551247360  8.381455e+17           2.195506e+07          
291   838085839343206401  8.380855e+17           2.894131e+09          
313   835246439529840640  8.352460e+17           2.625958e+07          
342   832088576586297345  8.320875e+17           3.058208e+07          
346   831926988323639298  8.319030e+17           2.068372e+07          
375   828361771580813312 NaN                    NaN                    
387   826598799820865537  8.265984e+17           4.196984e+09          
409   823333489516937216  8.233264e+17           1.582854e+09          
427   821153421864615936  8.211526e+17           1.132119e+08          
498   813130366689148928  8.131273e+17           4.196984e+09          
513   811647686436880384  8.116272e+17           4.196984e+09          
570   801854953262350336  8.018543e+17           1.185634e+07          
576   800859414831898624  8.008580e+17           2.918590e+08          
611   797165961484890113  7.971238e+17           2.916630e+07          
701   786051337297522688  7.727430e+17           7.305050e+17          
707   785515384317313025 NaN                    NaN                    
843   766714921925144576  7.667118e+17           4.196984e+09          
857   763956972077010945  7.638652e+17           1.584641e+07          
967   750381685133418496  7.501805e+17           4.717297e+09          
1005  747651430853525504  7.476487e+17           4.196984e+09          
1080  738891149612572673  7.384119e+17           3.589728e+08          
1295  707983188426153984  7.079801e+17           2.319108e+09          
1345  704491224099647488  7.044857e+17           2.878549e+07          
1445  696518437233913856 NaN                    NaN                    
1446  696490539101908992  6.964887e+17           4.196984e+09          
1474  693644216740769793  6.936422e+17           4.196984e+09          
1479  693582294167244802  6.935722e+17           1.198989e+09          
1497  692423280028966913  6.924173e+17           4.196984e+09          
1523  690607260360429569  6.903413e+17           4.670367e+08          
1598  686035780142297088  6.860340e+17           4.196984e+09          
1605  685681090388975616  6.855479e+17           4.196984e+09          
1618  684969860808454144  6.849598e+17           4.196984e+09          
1663  682808988178739200  6.827884e+17           4.196984e+09          
1689  681340665377193984  6.813394e+17           4.196984e+09          
1774  678023323247357953  6.780211e+17           4.196984e+09          
1819  676590572941893632  6.765883e+17           4.196984e+09          
1844  675849018447167488  6.758457e+17           4.196984e+09          
1895  674742531037511680  6.747400e+17           4.196984e+09          
1905  674606911342424069  6.744689e+17           4.196984e+09          
1914  674330906434379776  6.658147e+17           1.637468e+07          
1940  673716320723169284  6.737159e+17           4.19698

In [95]:
df_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [96]:
df_image_predictions

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
...                  ...   
2045  886366144734445568   
2046  886680336477933568   
2047  886736880519319552   
2048  886983233522544640   
2049  887101392804085760   
2050  887343217045368832   
2051  887473957103951883   
2052  887517139158093824   
2053  887705289381826560   
2054  888078434458587136   
2055  888202515573088257   
2056  888554962724278272   
2057  888804989199671297   
2058  888917238123831296   
2059  889278841981685760   
2060  889531135344209921   
2061  889638837579907072   
2062  889665388333682689   
2063  889880896479866881   
2064  890006608113172480   
2065  890240255349198849   
2066  890609185150312448   
2067  890729181411237888   
2068  890971913173991426   
2069  891087950875897856   
2070  891327558926688256   
2071  891689557279858688   
2072  891815181378084864   
2073  892177421306343426   
2074  892420643555336193   

                                                                                      jpg_url  \
0     https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg                                           
1     https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg                                           
2     https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg                                           
3     https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg                                           
4     https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg                                           
5     https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg                                           
6     https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg                                           
7     https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg                                           
8     https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg                                           
9     https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg                                           
10    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg                                           
11    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg                                           
12    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg                                           
13    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg                                           
14    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg                                           
15    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg                                           
16    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg                                           
17    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg                                           
18    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg                                           
19    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg                                           
20    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg                                           
21    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg                                           
22    https://pbs.twimg.com/media/CT9OwFIWEAMuRj

In [97]:
df_image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [98]:
df_tweet

tweet_id  retweet_count  favorite_count
0     892420643555336193  8315           38014         
1     892177421306343426  6142           32641         
2     891815181378084864  4067           24565         
3     891689557279858688  8450           41365         
4     891327558926688256  9157           39551         
5     891087950875897856  3046           19860         
6     890971913173991426  2019           11614         
7     890729181411237888  18445          64155         
8     890609185150312448  4183           27288         
9     890240255349198849  7215           31306         
10    890006608113172480  7172           30087         
11    889880896479866881  4869           27279         
12    889665388333682689  9828           47209         
13    889638837579907072  4436           26639         
14    889531135344209921  2198           14835         
15    889278841981685760  5254           24774         
16    888917238123831296  4400           28576         
17    888804989199671297  4193           25084         
18    888554962724278272  3472           19451         
19    888078434458587136  3415           21348         
20    887705289381826560  5275           29612         
21    887517139158093824  11426          45401         
22    887473957103951883  17776          67760         
23    887343217045368832  10176          33065         
24    887101392804085760  5837           30002         
25    886983233522544640  7603           34498         
26    886736880519319552  3211           11834         
27    886680336477933568  4365           22010         
28    886366144734445568  3130           20793         
29    886267009285017600  4              116           
...                  ... ..              ...           
2309  666411507551481857  323            438           
2310  666407126856765440  40             105           
2311  666396247373291520  84             164           
2312  666373753744588802  90             185           
2313  666362758909284353  562            763           
2314  666353288456101888  71             215           
2315  666345417576210432  133            288           
2316  666337882303524864  90             193           
2317  666293911632134144  348            490           
2318  666287406224695296  64             144           
2319  666273097616637952  76             171           
2320  666268910803644416  35             100           
2321  666104133288665088  6433           14081         
2322  666102155909144576  12             78            
2323  666099513787052032  66             150           
2324  666094000022159362  72             160           
2325  666082916733198337  44             114           
2326  666073100786774016  160            316           
2327  666071193221509120  59             142           
2328  666063827256086533  213            465           
2329  666058600524156928  57             109           
2330  666057090499244032  139            290           
2331  666055525042405380  238            429           
2332  666051853826850816  837            1203          
2333  666050758794694657  59             130           
2334  666049248165822465  41             106           
2335  666044226329800704  139            292           
2336  666033412701032449  43             123           
2337  666029285002620928  47             126           
2338  666020888022790149  502            2535          

[2339 rows x 3 columns]

In [99]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2339 entries, 0 to 2338
Data columns (total 3 columns):
tweet_id          2339 non-null int64
retweet_count     2339 non-null int64
favorite_count    2339 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


### Quality 
#### `df_twitter_archive` table
- 181 Retweets ( 78 replies??)
- Missing data in expanded_urls
- (Tweets without images) ???
- The ratings aren't all correct
- Incorrect dog names (A, an, the, etc.)
- Missing values in dog names and dog stages. (represented as None)
- Erroneous datatype (timestamp, retweeted_status_timestamp, tweet_id, in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, .???.)
- Not all images are dog images
- Hard to understand the source from source column
- Dog stage's type to categorical

#### `df_image_predictions` table
- Missing records (2075 instead of 2356, some of them are probably the retweets or with no images)
- Image predictions that are not dog breed
- Lowercase breed names in p1, p2, p3


### Tidiness
- Merge three data frames.
- Drop unneeded columns.
- Combine dog "stage" columns (i.e. doggo, floofer, pupper, and puppo) into one column.


## Clean Data

In [100]:
# Create the copies of the dataframes which will be cleaned
df_twitter_archive_clean = df_twitter_archive.copy()
df_image_predictions_clean = df_image_predictions.copy()
df_tweet_clean = df_tweet.copy()

#### Remove Retweets

##### Define

Remove 181 columns which are retweets and detected by the columns:retweeted_status_id, retweeted_status_user_id and retweeted_status_timestamp. Remove retweets by keeping the null rows of retweeted_status_id column.

##### Code

In [101]:
df_twitter_archive_clean = df_twitter_archive_clean[df_twitter_archive_clean.retweeted_status_id.isnull()]

##### Test

In [102]:
 df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

#### Missing data in expanded_urls

##### Define

Remove the tweets without expanded_urls i.e images by keeping the non null data in the expanded_urls column.

##### Code

In [103]:
df_twitter_archive_clean = df_twitter_archive_clean[df_twitter_archive_clean.expanded_urls.notnull()]


##### Test

In [104]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2117 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2117 non-null object
source                        2117 non-null object
text                          2117 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2117 non-null int64
rating_denominator            2117 non-null int64
name                          2117 non-null object
doggo                         2117 non-null object
floofer                       2117 non-null object
pupper                        2117 non-null object
puppo                         2117 non-null object
dtypes: float64(4), int64(3), object(1

#### Merge three data frames

##### Define

Merge the `df_tweet_clean` and `df_image_predictions` tables to the `df_twitter_archive_clean` table, both joining on *tweet_id*.

##### Code

In [105]:
# First, merge df_tweet_clean table to the df_twitter_archive_clean
df_twitter_archive_clean = pd.merge(df_twitter_archive_clean, df_tweet_clean, 
                                    on = ['tweet_id'], how = 'left')

##### Test

In [106]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2116
Data columns (total 19 columns):
tweet_id                      2117 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2117 non-null object
source                        2117 non-null object
text                          2117 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2117 non-null int64
rating_denominator            2117 non-null int64
name                          2117 non-null object
doggo                         2117 non-null object
floofer                       2117 non-null object
pupper                        2117 non-null object
puppo                         2117 non-null object
retweet_count                 2114 non

In [107]:
# Second, merge df_image_predictions table to the df_twitter_archive_clean
df_twitter_archive_clean = pd.merge(df_twitter_archive_clean, df_image_predictions, 
                                    on = ['tweet_id'], how = 'left')

In [108]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2116
Data columns (total 30 columns):
tweet_id                      2117 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2117 non-null object
source                        2117 non-null object
text                          2117 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2117 non-null int64
rating_denominator            2117 non-null int64
name                          2117 non-null object
doggo                         2117 non-null object
floofer                       2117 non-null object
pupper                        2117 non-null object
puppo                         2117 non-null object
retweet_count                 2114 non

#### Drop unneeded columns

##### Define

Drop the the columns related to the retweets and the replies: *retweeted_status_id*, *retweeted_status_user_id*, *retweeted_status_timestamp*, *in_reply_to_status_id*, *in_reply_to_user_id*

##### Code

In [109]:
df_twitter_archive_clean = df_twitter_archive_clean.drop(['retweeted_status_id','retweeted_status_user_id',
                              'retweeted_status_timestamp', 'in_reply_to_status_id', 'in_reply_to_user_id'], axis=1)

##### Test

In [110]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2116
Data columns (total 25 columns):
tweet_id              2117 non-null int64
timestamp             2117 non-null object
source                2117 non-null object
text                  2117 non-null object
expanded_urls         2117 non-null object
rating_numerator      2117 non-null int64
rating_denominator    2117 non-null int64
name                  2117 non-null object
doggo                 2117 non-null object
floofer               2117 non-null object
pupper                2117 non-null object
puppo                 2117 non-null object
retweet_count         2114 non-null float64
favorite_count        2114 non-null float64
jpg_url               1994 non-null object
img_num               1994 non-null float64
p1                    1994 non-null object
p1_conf               1994 non-null float64
p1_dog                1994 non-null object
p2                    1994 non-null object
p2_conf               1994 non-n

#### Combine dog "stage" columns (i.e. doggo, floofer, pupper, and puppo) into one column.


##### Define

Create a new column named *dog_stage*. Extract doggo, floofer, pupper and puppo from the *text* column using pandas' `str.extact method` and place them into the *dog_stage* column. Fill the null values with None.

##### Code

In [111]:
df_twitter_archive_clean['dog_stage'] = df_twitter_archive_clean['text'].str.extract('(doggo|floofer|pupper|puppo)',
                                                                             expand=True)
df_twitter_archive_clean = df_twitter_archive_clean.drop(['doggo','floofer','pupper','puppo'], axis=1)
df_twitter_archive_clean['dog_stage'].fillna('None', inplace=True)

##### Test

In [112]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2116
Data columns (total 22 columns):
tweet_id              2117 non-null int64
timestamp             2117 non-null object
source                2117 non-null object
text                  2117 non-null object
expanded_urls         2117 non-null object
rating_numerator      2117 non-null int64
rating_denominator    2117 non-null int64
name                  2117 non-null object
retweet_count         2114 non-null float64
favorite_count        2114 non-null float64
jpg_url               1994 non-null object
img_num               1994 non-null float64
p1                    1994 non-null object
p1_conf               1994 non-null float64
p1_dog                1994 non-null object
p2                    1994 non-null object
p2_conf               1994 non-null float64
p2_dog                1994 non-null object
p3                    1994 non-null object
p3_conf               1994 non-null float64
p3_dog                1994 non

In [113]:
df_twitter_archive_clean['dog_stage']

0       None  
1       None  
2       None  
3       None  
4       None  
5       None  
6       None  
7       None  
8       None  
9       doggo 
10      None  
11      None  
12      puppo 
13      None  
14      puppo 
15      None  
16      None  
17      None  
18      None  
19      None  
20      None  
21      None  
22      None  
23      None  
24      None  
25      None  
26      None  
27      None  
28      pupper
29      None  
        ...   
2087    None  
2088    None  
2089    None  
2090    None  
2091    None  
2092    None  
2093    None  
2094    None  
2095    None  
2096    None  
2097    None  
2098    None  
2099    None  
2100    None  
2101    None  
2102    None  
2103    None  
2104    None  
2105    None  
2106    None  
2107    None  
2108    None  
2109    None  
2110    None  
2111    None  
2112    None  
2113    None  
2114    None  
2115    None  
2116    None  
Name: dog_stage, Length: 2117, dtype: object

In [114]:
df_twitter_archive_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name  retweet_count  \
0  13                10                  Phineas   8315.0          
1  13                10                  Tilly     6142.0          
2  12                10                  Archie    4067.0          
3  13                10                  Darla     8450.0          
4  12                10                  Franklin  9157.0          

   favorite_count    ...               p1   p1_conf p1_dog  \
0  38014.0           ...      orange       0.097049  False   
1  32641.0           ...      Chihuahua    0.323581  True    
2  24565.0           ...      Chihuahua    0.716012  True    
3  41365.0           ...      paper_towel  0.170278  False   
4  39551.0           ...      basset       0.555712  True    

                   p2   p2_conf p2_dog                           p3   p3_conf  \
0  bagel               0.085851  False  banana                       0.076110   
1  Pekinese            0.090647  True   papillon                     0.068957   
2  malamute            0.078253  True   kelpie                       0.031379   
3  Labrador_retriever  0.168086  True   spatula                      0.040836   
4  English_springer    0.225770  True   German_short-haired_pointer  0.175219   

  p3_dog  dog_stage  
0  False  None       
1  True   None       
2  True   

#### Not all images are dog images


##### Define

Find the images that do not contain dogs by exploring the ratings and the text, then drop those data from the `df_twitter_archive_clean` data frame. 

##### Code

In [115]:
# rating_numerator is usually greater than 10, so tweets with numerator of 0 should be checked
df_twitter_archive_clean.query('rating_numerator == 0')[['text','expanded_urls']]

text  \
249  When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag         
807  PUPDATE: can't see any. Even if I could, I couldn't reach them to pet. 0/10 much disappointment https://t.co/c7WXaB2nqX   

                                                                                                                                                                                       expanded_urls  
249  https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1  
807  https://twitter.com/dog_rates/status/746906459439529985/photo/1

These two images with the 0 rating_numerator are not dogs' images and should be dropped.

In [116]:
df_twitter_archive_clean = df_twitter_archive_clean.drop(labels=[249,807],axis=0)

In [117]:
# Test if the rows are dropped
df_twitter_archive_clean.query('rating_numerator == 0')[['text','expanded_urls']]

Empty DataFrame
Columns: [text, expanded_urls]
Index: []

In case of getting an image of something else rather than dogs the following statements are written in the text: 'only rate dogs','only send in dogs','only send dogs', 'stop sending', 'without a dog','not pet'

In [118]:
no_dogs = 'only rate dogs|only send in dogs|only send dogs|stop sending|without a dog|not pet'
df_twitter_archive_clean[df_twitter_archive_clean['text'].str.contains(no_dogs)]

tweet_id                  timestamp  \
24    887101392804085760  2017-07-18 00:07:08 +0000   
43    883117836046086144  2017-07-07 00:17:54 +0000   
54    880872448815771648  2017-06-30 19:35:32 +0000   
79    875021211251597312  2017-06-14 16:04:48 +0000   
82    874057562936811520  2017-06-12 00:15:36 +0000   
97    870804317367881728  2017-06-03 00:48:22 +0000   
109   867900495410671616  2017-05-26 00:29:37 +0000   
112   867051520902168576  2017-05-23 16:16:06 +0000   
120   864873206498414592  2017-05-17 16:00:15 +0000   
131   862096992088072192  2017-05-10 00:08:34 +0000   
142   859196978902773760  2017-05-02 00:04:57 +0000   
158   855459453768019968  2017-04-21 16:33:22 +0000   
163   854010172552949760  2017-04-17 16:34:26 +0000   
180   849776966551130114  2017-04-06 00:13:11 +0000   
182   849336543269576704  2017-04-04 19:03:06 +0000   
184   848690551926992896  2017-04-03 00:16:10 +0000   
192   847116187444137987  2017-03-29 16:00:12 +0000   
200   845677943972139009  2017-03-25 16:45:08 +0000   
248   835172783151792128  2017-02-24 17:01:22 +0000   
309   826240494070030336  2017-01-31 01:27:39 +0000   
329   822610361945911296  2017-01-21 01:02:48 +0000   
332   822244816520155136  2017-01-20 00:50:15 +0000   
386   813187593374461952  2016-12-26 01:00:05 +0000   
410   809920764300447744  2016-12-17 00:38:52 +0000   
426   806219024703037440  2016-12-06 19:29:28 +0000   
563   781524693396357120  2016-09-29 16:03:01 +0000   
573   780192070812196864  2016-09-25 23:47:39 +0000   
618   772581559778025472  2016-09-04 23:46:12 +0000   
630   771014301343748096  2016-08-31 15:58:28 +0000   
633   770655142660169732  2016-08-30 16:11:18 +0000   
...                  ...                        ...   
938   726224900189511680  2016-04-30 01:41:23 +0000   
981   717537687239008257  2016-04-06 02:21:30 +0000   
995   715733265223708672  2016-04-01 02:51:22 +0000   
1007  714631576617938945  2016-03-29 01:53:39 +0000   
1013  714141408463036416  2016-03-27 17:25:54 +0000   
1022  712717840512598017  2016-03-23 19:09:09 +0000   
1046  710272297844797440  2016-03-17 01:11:26 +0000   
1077  708130923141795840  2016-03-11 03:22:23 +0000   
1141  703425003149250560  2016-02-27 03:42:44 +0000   
1147  703041949650034688  2016-02-26 02:20:37 +0000   
1157  702217446468493312  2016-02-23 19:44:20 +0000   
1170  700747788515020802  2016-02-19 18:24:26 +0000   
1220  697259378236399616  2016-02-10 03:22:44 +0000   
1244  694925794720792577  2016-02-03 16:49:55 +0000   
1256  693647888581312512  2016-01-31 04:11:58 +0000   
1269  692919143163629568  2016-01-29 03:56:12 +0000   
1306  690360449368465409  2016-01-22 02:28:52 +0000   
1343  688116655151435777  2016-01-15 21:52:49 +0000   
1387  685532292383666176  2016-01-08 18:43:29 +0000   
1467  681297372102656000  2015-12-28 02:15:26 +0000   
1473  680970795137544192  2015-12-27 04:37:44 +0000   
1500  680055455951884288  2015-12-24 16:00:30 +0000   
1511  679530280114372609  2015-12-23 05:13:38 +0000   
1519  679148763231985668  2015-12-22 03:57:37 +0000   
1551  677895101218201600  2015-12-18 16:56:01 +0000   
1613  675878199931371520  2015-12-13 03:21:34 +0000   
1625  675534494439489536  2015-12-12 04:35:48 +0000   
1640  675153376133427200  2015-12-11 03:21:23 +0000   
1648  675109292475830276  2015-12-11 00:26:12 +0000   
1706  673906403526995968  2015-12-07 16:46:21 +0000   

                                                                                   source  \
24    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
43    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
54    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
79    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
82    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
97    <a href="http://twitter.com/download/ipho

In [119]:
df_twitter_archive_clean = df_twitter_archive_clean[~df_twitter_archive_clean['text'].str.contains(no_dogs)]


##### Test

In [120]:
df_twitter_archive_clean[df_twitter_archive_clean['text'].str.contains(no_dogs)]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, dog_stage]
Index: []

[0 rows x 22 columns]

In [121]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2038 entries, 0 to 2116
Data columns (total 22 columns):
tweet_id              2038 non-null int64
timestamp             2038 non-null object
source                2038 non-null object
text                  2038 non-null object
expanded_urls         2038 non-null object
rating_numerator      2038 non-null int64
rating_denominator    2038 non-null int64
name                  2038 non-null object
retweet_count         2036 non-null float64
favorite_count        2036 non-null float64
jpg_url               1916 non-null object
img_num               1916 non-null float64
p1                    1916 non-null object
p1_conf               1916 non-null float64
p1_dog                1916 non-null object
p2                    1916 non-null object
p2_conf               1916 non-null float64
p2_dog                1916 non-null object
p3                    1916 non-null object
p3_conf               1916 non-null float64
p3_dog                1916 non

#### Incorrect dog names (a, an, the, etc.)

##### Define

Names are probably extracted after the statements 'This is','Here is', 'hello to', 'meet'. That is why we have lots of words in the name column such that; a, an, the, etc. Notice that all of those are lowercase. Find the lowercase names in *name* cloumn and re-extract the names that come after the statements: 'name is' and 'named'. Replace the rest of the lowercase names with 'None'. Also, change the name 'O' with "O'Malley".

##### Code

In [122]:
df_twitter_archive_clean[df_twitter_archive_clean['name'].str.len() == 1]

tweet_id                  timestamp  \
51    881536004380872706  2017-07-02 15:32:16 +0000   
497   792913359805018113  2016-10-31 02:17:31 +0000   
596   776201521193218049  2016-09-14 23:30:38 +0000   
795   747885874273214464  2016-06-28 20:14:22 +0000   
797   747816857231626240  2016-06-28 15:40:07 +0000   
838   743222593470234624  2016-06-15 23:24:09 +0000   
1126  704859558691414016  2016-03-02 02:43:09 +0000   
1136  704054845121142784  2016-02-28 21:25:30 +0000   
1146  703079050210877440  2016-02-26 04:48:02 +0000   
1153  702539513671897089  2016-02-24 17:04:07 +0000   
1167  700864154249383937  2016-02-20 02:06:50 +0000   
1279  692187005137076224  2016-01-27 03:26:56 +0000   
1558  677644091929329666  2015-12-18 00:18:36 +0000   
1624  675706639471788032  2015-12-12 15:59:51 +0000   
1649  675047298674663426  2015-12-10 20:19:52 +0000   
1691  674082852460433408  2015-12-08 04:27:30 +0000   
1708  673715861853720576  2015-12-07 04:09:13 +0000   
1722  673636718965334016  2015-12-06 22:54:44 +0000   
1761  672604026190569472  2015-12-04 02:31:10 +0000   
1801  671743150407421952  2015-12-01 17:30:22 +0000   
1832  671147085991960577  2015-11-30 02:01:49 +0000   
1882  670427002554466305  2015-11-28 02:20:27 +0000   
1891  670361874861563904  2015-11-27 22:01:40 +0000   
1894  670303360680108032  2015-11-27 18:09:09 +0000   
1912  669923323644657664  2015-11-26 16:59:01 +0000   
1918  669661792646373376  2015-11-25 23:39:47 +0000   
1926  669564461267722241  2015-11-25 17:13:02 +0000   
1955  668955713004314625  2015-11-24 00:54:05 +0000   
1962  668815180734689280  2015-11-23 15:35:39 +0000   
1975  668614819948453888  2015-11-23 02:19:29 +0000   
1982  668507509523615744  2015-11-22 19:13:05 +0000   
1986  668466899341221888  2015-11-22 16:31:42 +0000   
1999  668171859951755264  2015-11-21 20:59:20 +0000   
2013  667861340749471744  2015-11-21 00:25:26 +0000   
2019  667773195014021121  2015-11-20 18:35:10 +0000   
2026  667538891197542400  2015-11-20 03:04:08 +0000   
2035  667470559035432960  2015-11-19 22:32:36 +0000   
2049  667177989038297088  2015-11-19 03:10:02 +0000   
2065  666983947667116034  2015-11-18 14:18:59 +0000   
2072  666781792255496192  2015-11-18 00:55:42 +0000   
2075  666701168228331520  2015-11-17 19:35:19 +0000   
2088  666407126856765440  2015-11-17 00:06:54 +0000   
2095  666293911632134144  2015-11-16 16:37:02 +0000   
2108  666057090499244032  2015-11-16 00:55:59 +0000   
2109  666055525042405380  2015-11-16 00:49:46 +0000   
2111  666050758794694657  2015-11-16 00:30:50 +0000   
2113  666044226329800704  2015-11-16 00:04:52 +0000   
2114  666033412701032449  2015-11-15 23:21:54 +0000   
2115  666029285002620928  2015-11-15 23:05:30 +0000   

                                                                                  source  \
51    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
497   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
596   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
795   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
797   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
838   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1126  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1136  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1146  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1153  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1167  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1279  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1558  <a href="http://twitter.com/download/iphone" rel="nofollow">Twi

In [123]:
df_twitter_archive_clean.query('name == "a"')['text'].count()

48

In [124]:
pd.set_option('display.max_colwidth',-1)

df_twitter_archive_clean['name'] = df_twitter_archive_clean['name'].str.replace('O', "O'Malley")

In [125]:
df_twitter_archive_clean.query('name == "a"')['text']

51      Here is a pupper approaching maximum borkdrive. Zooming at never before seen speeds. 14/10 paw-inspiring af \n(IG: puffie_the_chow) https://t.co/ghXBIIeQZF
497     Here is a perfect example of someone who has their priorities in order. 13/10 for both owner and Forrest https://t.co/LRyMrU7Wfq                           
795     This is a mighty rare blue-tailed hammer sherk. Human almost lost a limb trying to take these. Be careful guys. 8/10 https://t.co/TGenMeXreW               
797     Viewer discretion is advised. This is a terrible attack in progress. Not even in water (tragic af). 4/10 bad sherk https://t.co/L3U0j14N5R                 
838     This is a very rare Great Alaskan Bush Pupper. Hard to stumble upon without spooking. 12/10 would pet passionately https://t.co/xOBKCdpzaa                 
1126    Here is a heartbreaking scene of an incredible pupper being laid to rest. 10/10 RIP pupper https://t.co/81mvJ0rGRu                                         
1136    Here is 

In [126]:
df_twitter_archive_clean.loc[df_twitter_archive_clean['name'].str.islower()]

tweet_id                  timestamp  \
21    887517139158093824  2017-07-19 03:39:09 +0000   
51    881536004380872706  2017-07-02 15:32:16 +0000   
266   832645525019123713  2017-02-17 17:38:57 +0000   
292   828650029636317184  2017-02-06 17:02:17 +0000   
497   792913359805018113  2016-10-31 02:17:31 +0000   
659   765395769549590528  2016-08-16 03:52:26 +0000   
722   755206590534418437  2016-07-19 01:04:16 +0000   
785   748692773788876800  2016-07-01 01:40:41 +0000   
786   748575535303884801  2016-06-30 17:54:50 +0000   
795   747885874273214464  2016-06-28 20:14:22 +0000   
797   747816857231626240  2016-06-28 15:40:07 +0000   
830   744223424764059648  2016-06-18 17:41:06 +0000   
838   743222593470234624  2016-06-15 23:24:09 +0000   
852   741067306818797568  2016-06-10 00:39:48 +0000   
883   736392552031657984  2016-05-28 03:04:00 +0000   
908   731156023742988288  2016-05-13 16:15:54 +0000   
926   728035342121635841  2016-05-05 01:35:26 +0000   
994   715758151270801409  2016-04-01 04:30:16 +0000   
1126  704859558691414016  2016-03-02 02:43:09 +0000   
1136  704054845121142784  2016-02-28 21:25:30 +0000   
1146  703079050210877440  2016-02-26 04:48:02 +0000   
1153  702539513671897089  2016-02-24 17:04:07 +0000   
1167  700864154249383937  2016-02-20 02:06:50 +0000   
1240  695095422348574720  2016-02-04 04:03:57 +0000   
1279  692187005137076224  2016-01-27 03:26:56 +0000   
1381  685943807276412928  2016-01-09 21:58:42 +0000   
1498  680085611152338944  2015-12-24 18:00:19 +0000   
1521  679111216690831360  2015-12-22 01:28:25 +0000   
1558  677644091929329666  2015-12-18 00:18:36 +0000   
1570  677269281705472000  2015-12-16 23:29:14 +0000   
...                  ...                        ...   
1955  668955713004314625  2015-11-24 00:54:05 +0000   
1962  668815180734689280  2015-11-23 15:35:39 +0000   
1968  668636665813057536  2015-11-23 03:46:18 +0000   
1975  668614819948453888  2015-11-23 02:19:29 +0000   
1976  668587383441514497  2015-11-23 00:30:28 +0000   
1982  668507509523615744  2015-11-22 19:13:05 +0000   
1986  668466899341221888  2015-11-22 16:31:42 +0000   
1999  668171859951755264  2015-11-21 20:59:20 +0000   
2013  667861340749471744  2015-11-21 00:25:26 +0000   
2019  667773195014021121  2015-11-20 18:35:10 +0000   
2026  667538891197542400  2015-11-20 03:04:08 +0000   
2035  667470559035432960  2015-11-19 22:32:36 +0000   
2049  667177989038297088  2015-11-19 03:10:02 +0000   
2065  666983947667116034  2015-11-18 14:18:59 +0000   
2072  666781792255496192  2015-11-18 00:55:42 +0000   
2075  666701168228331520  2015-11-17 19:35:19 +0000   
2087  666411507551481857  2015-11-17 00:24:19 +0000   
2088  666407126856765440  2015-11-17 00:06:54 +0000   
2094  666337882303524864  2015-11-16 19:31:45 +0000   
2095  666293911632134144  2015-11-16 16:37:02 +0000   
2096  666287406224695296  2015-11-16 16:11:11 +0000   
2106  666063827256086533  2015-11-16 01:22:45 +0000   
2107  666058600524156928  2015-11-16 01:01:59 +0000   
2108  666057090499244032  2015-11-16 00:55:59 +0000   
2109  666055525042405380  2015-11-16 00:49:46 +0000   
2110  666051853826850816  2015-11-16 00:35:11 +0000   
2111  666050758794694657  2015-11-16 00:30:50 +0000   
2113  666044226329800704  2015-11-16 00:04:52 +0000   
2114  666033412701032449  2015-11-15 23:21:54 +0000   
2115  666029285002620928  2015-11-15 23:05:30 +0000   

                                                                                   source  \
21    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
51    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
266   <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                    
292   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
497   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
659   <a href="http://twitter.com/download/ipho

In [127]:
df_twitter_archive_clean.query('name == "a"')['text'].count()

48

In [128]:
df_changed_names = df_twitter_archive_clean[df_twitter_archive_clean['name'].str.islower()]
for text in df_changed_names.text:
    found = re.findall(r'(?:named|name is)\s(\w+)', text)
    mask = df_twitter_archive_clean['text'] == text
    if len(found) > 0:
        df_twitter_archive_clean.loc[mask, 'name'] = found[0]
    else:
        df_twitter_archive_clean.loc[mask, 'name'] = 'None'
    

##### Test

In [129]:
df_twitter_archive_clean.query('name == "a"')['text'].count()

0

In [130]:
df_twitter_archive_clean

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
19    888078434458587136  2017-07-20 16:49:33 +0000   
20    887705289381826560  2017-07-19 16:06:48 +0000   
21    887517139158093824  2017-07-19 03:39:09 +0000   
22    887473957103951883  2017-07-19 00:47:34 +0000   
23    887343217045368832  2017-07-18 16:08:03 +0000   
25    886983233522544640  2017-07-17 16:17:36 +0000   
26    886736880519319552  2017-07-16 23:58:41 +0000   
27    886680336477933568  2017-07-16 20:14:00 +0000   
28    886366144734445568  2017-07-15 23:25:31 +0000   
29    886258384151887873  2017-07-15 16:17:19 +0000   
30    885984800019947520  2017-07-14 22:10:11 +0000   
...                  ...                        ...   
2087  666411507551481857  2015-11-17 00:24:19 +0000   
2088  666407126856765440  2015-11-17 00:06:54 +0000   
2089  666396247373291520  2015-11-16 23:23:41 +0000   
2090  666373753744588802  2015-11-16 21:54:18 +0000   
2091  666362758909284353  2015-11-16 21:10:36 +0000   
2092  666353288456101888  2015-11-16 20:32:58 +0000   
2093  666345417576210432  2015-11-16 20:01:42 +0000   
2094  666337882303524864  2015-11-16 19:31:45 +0000   
2095  666293911632134144  2015-11-16 16:37:02 +0000   
2096  666287406224695296  2015-11-16 16:11:11 +0000   
2097  666273097616637952  2015-11-16 15:14:19 +0000   
2098  666268910803644416  2015-11-16 14:57:41 +0000   
2099  666104133288665088  2015-11-16 04:02:55 +0000   
2100  666102155909144576  2015-11-16 03:55:04 +0000   
2101  666099513787052032  2015-11-16 03:44:34 +0000   
2102  666094000022159362  2015-11-16 03:22:39 +0000   
2103  666082916733198337  2015-11-16 02:38:37 +0000   
2104  666073100786774016  2015-11-16 01:59:36 +0000   
2105  666071193221509120  2015-11-16 01:52:02 +0000   
2106  666063827256086533  2015-11-16 01:22:45 +0000   
2107  666058600524156928  2015-11-16 01:01:59 +0000   
2108  666057090499244032  2015-11-16 00:55:59 +0000   
2109  666055525042405380  2015-11-16 00:49:46 +0000   
2110  666051853826850816  2015-11-16 00:35:11 +0000   
2111  666050758794694657  2015-11-16 00:30:50 +0000   
2112  666049248165822465  2015-11-16 00:24:50 +0000   
2113  666044226329800704  2015-11-16 00:04:52 +0000   
2114  666033412701032449  2015-11-15 23:21:54 +0000   
2115  666029285002620928  2015-11-15 23:05:30 +0000   
2116  666020888022790149  2015-11-15 22:32:08 +0000   

                                                                                  source  \
0     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
5     <a href="http://twitter.com/download/iphone" re

#### The ratings aren't all correct

##### Define

The rating_denominator is generally 10 or the multiples of 10. Find the expections and look if they are correctly extracted. Re-extract the tweets that have decimal rating_numerator (5 is extracted instead of 13.5). Some tweets have multiple ratings because their images contain multiple dogs. Also, in some tweets there are puppies and that is why we have the rating of 60/50. 

##### Code

In [131]:
df_twitter_archive_clean.query('rating_numerator == 1')

tweet_id                  timestamp  \
1857  670783437142401025  2015-11-29 01:56:48 +0000   
2023  667549055577362432  2015-11-20 03:44:31 +0000   
2096  666287406224695296  2015-11-16 16:11:11 +0000   
2099  666104133288665088  2015-11-16 04:02:55 +0000   

                                                                                  source  \
1857  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2023  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   
2096  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2099  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
1857  Flamboyant pup here. Probably poisonous. Won't eat kibble. Doesn't bark. Slow af. Petting doesn't look fun. 1/10 https://t.co/jxukeh2BeO       
2023  Never seen dog like this. Breathes heavy. Tilts head in a pattern. No bark. Shitty at fetch. Not even cordless. 1/10 https://t.co/i9iSGNn3fx   
2096  This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv       
2099  Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/Asgdc6kuLX         

                                                        expanded_urls  \
1857  https://twitter.com/dog_rates/status/670783437142401025/photo/1   
2023  https://twitter.com/dog_rates/status/667549055577362432/photo/1   
2096  https://twitter.com/dog_rates/status/666287406224695296/photo/1   
2099  https://twitter.com/dog_rates/status/666104133288665088/photo/1   

      rating_numerator  rating_denominator  name  retweet_count  \
1857  1                 10                  None  406.0           
2023  1                 10                  None  2323.0          
2096  1                 2                   None  64.0            
2099  1                 10                  None  6433.0          

      favorite_count    ...                p1   p1_conf p1_dog  \
1857  831.0             ...      lacewing      0.381955  False   
2023  5841.0            ...      electric_fan  0.984377  False   
2096  144.0             ...      Maltese_dog   0.857531  True    
2099  14081.0           ...      hen           0.965932  False   

                     p2   p2_conf p2_dog                p3   p3_conf p3_dog  \
1857  sulphur_butterfly  0.106810  False  leafhopper        0.068347  False   
2023  spotlight          0.007737  False  lampshade         0.001901  False   
2096  toy_poodle         0.063064  True   miniature_poodle  0.025581  True    
2099  cock               0.033919  False  partridge         0.000052  False   

      dog_stage  
1857  None       
2023  None       
2096  None       
2099  None       

[4 rows x 22 columns]

In [132]:
# Find the tweet with decimal numerator
df_twitter_archive_clean[df_twitter_archive_clean.rating_numerator==5]

tweet_id                  timestamp  \
41    883482846933004288  2017-07-08 00:28:19 +0000   
561   781661882474196992  2016-09-30 01:08:10 +0000   
1184  699691744225525762  2016-02-16 20:28:06 +0000   
1288  691483041324204033  2016-01-25 04:49:38 +0000   
1362  687102708889812993  2016-01-13 02:43:46 +0000   
1395  684959798585110529  2016-01-07 04:48:36 +0000   
1400  684880619965411328  2016-01-06 23:33:58 +0000   
1421  683849932751646720  2016-01-04 03:18:23 +0000   
1455  682003177596559360  2015-12-30 01:00:03 +0000   
1501  679877062409191424  2015-12-24 04:11:37 +0000   
1569  677301033169788928  2015-12-17 01:35:24 +0000   
1581  676897532954456065  2015-12-15 22:52:02 +0000   
1592  676588346097852417  2015-12-15 02:23:26 +0000   
1632  675483430902214656  2015-12-12 01:12:54 +0000   
1645  675135153782571009  2015-12-11 02:08:58 +0000   
1671  674646392044941312  2015-12-09 17:46:48 +0000   
1674  674632714662858753  2015-12-09 16:52:27 +0000   
1693  674063288070742018  2015-12-08 03:09:46 +0000   
1746  672980819271634944  2015-12-05 03:28:25 +0000   
1780  672231046314901505  2015-12-03 01:49:05 +0000   
1793  671879137494245376  2015-12-02 02:30:43 +0000   
1829  671159727754231808  2015-11-30 02:52:03 +0000   
1858  670782429121134593  2015-11-29 01:52:48 +0000   
1875  670449342516494336  2015-11-28 03:49:14 +0000   
1900  670069087419133954  2015-11-27 02:38:14 +0000   
1905  670037189829525505  2015-11-27 00:31:29 +0000   
1918  669661792646373376  2015-11-25 23:39:47 +0000   
1946  668994913074286592  2015-11-24 03:29:51 +0000   
1970  668631377374486528  2015-11-23 03:25:17 +0000   
2006  667911425562669056  2015-11-21 03:44:27 +0000   
2073  666776908487630848  2015-11-18 00:36:17 +0000   
2112  666049248165822465  2015-11-16 00:24:50 +0000   

                                                                                  source  \
41    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
561   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1184  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1288  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1362  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1395  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1400  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1421  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1455  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1501  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1569  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1581  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1592  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1632  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1645  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1671  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1674  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1693  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1746  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1780  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1793  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1829  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1858  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone

In [133]:
# See the whole text
pd.set_option('display.max_colwidth',-1)
# Look at the data whose rating_denominator is different than 10.
df_twitter_archive_clean[df_twitter_archive_clean.rating_denominator!=10].text

341     The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                      
405     Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx
702     Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                           
857     After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ                             
908     Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                
953     Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                      

Notice 24/7 is not a rating so this tweet shoul be dropped.

In [134]:
df_twitter_archive_clean = df_twitter_archive_clean.drop(labels=405,axis=0)

In [135]:
# Check if it is dropped
df_twitter_archive_clean[df_twitter_archive_clean.rating_numerator==24]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, dog_stage]
Index: []

[0 rows x 22 columns]

Change the rating 50/50 with the correct rating 11/10, and 4/20 with 13/10

In [136]:
df_twitter_archive_clean=df_twitter_archive_clean.set_value(990, 'rating_numerator', '11')
df_twitter_archive_clean=df_twitter_archive_clean.set_value(990, 'rating_denominator', '10')
df_twitter_archive_clean=df_twitter_archive_clean.set_value(953, 'rating_numerator', '13')
df_twitter_archive_clean=df_twitter_archive_clean.set_value(953, 'rating_denominator', '10')


(df_twitter_archive_clean.rating_numerator.loc[990], df_twitter_archive_clean.rating_denominator.loc[990], 
df_twitter_archive_clean.rating_numerator.loc[953], df_twitter_archive_clean.rating_denominator.loc[953])


/Users/yagmur/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:1: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  if __name__ == '__main__':
/Users/yagmur/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:2: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  from ipykernel import kernelapp as app
/Users/yagmur/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:3: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  app.launch_new_instance()
/Users/yagmur/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead


(11, 10, 13, 10)

Re-extract the ratings by taking the rating that has a denominator of 10 or multiples of 10. Also extract the decimal numerators.

In [137]:
# extract the rating_numerator
numerator_new = df_twitter_archive_clean.text.str.extract(r'(\d+(?:\.\d)?)\/\d+0', expand=False)

df_twitter_archive_clean['rating_numerator_new'] = numerator_new

# Compare the old rating numerator and the new one (for debugging purposes).
old = list(df_twitter_archive_clean.rating_numerator)
text = list(df_twitter_archive_clean.text)
for index, rating in enumerate(numerator_new):
    rating_old = old[index]
    if str(rating) != str(rating_old):
        print("Old: " + str(rating_old) + " New: " + str(rating) + "\n")
        print(text[index])

Old: 5 New: 13.5

This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948
Old: 9 New: 14

After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ
Old: 13 New: 4

Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a
Old: 11 New: 50

This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq
Old: 7 New: 10

This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5
Old: 1 New: 9

This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv


In [138]:
# extract the rating_denominator
denominator_new = df_twitter_archive_clean.text.str.extract(r'\d+(?:\.\d)?\/(\d+0)', expand=False)
df_twitter_archive_clean['rating_denominator_new'] = denominator_new

# Compare the old rating numerator and the new one.
old = list(df_twitter_archive_clean.rating_denominator)
text = list(df_twitter_archive_clean.text)
for index, rating in enumerate(denominator_new):
    denominator_old = old[index]
    if str(rating) != str(denominator_old):
        print("Old: " + str(denominator_old) + " New: " + str(rating) + "\n")
        print(text[index])

Old: 11 New: 10

After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ
Old: 10 New: 20

Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a
Old: 10 New: 50

This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq
Old: 11 New: 10

This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5
Old: 2 New: 10

This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv


In [139]:
df_twitter_archive_clean['rating_denominator_new'] = denominator_new

In [140]:
df_twitter_archive_clean.rating_numerator_new = df_twitter_archive_clean.rating_numerator_new.astype(float)
df_twitter_archive_clean.rating_denominator_new = df_twitter_archive_clean.rating_denominator_new.astype(float)


In [141]:
# Drop the old rating_numerator and rating_denominator columnns.
df_twitter_archive_clean = df_twitter_archive_clean.drop(['rating_numerator','rating_denominator'], axis=1)

##### Test

In [142]:
df_twitter_archive_clean[df_twitter_archive_clean.rating_denominator_new != 10]


tweet_id                  timestamp  \
341   820690176645140481  2017-01-15 17:52:40 +0000   
702   758467244762497024  2016-07-28 01:00:57 +0000   
908   731156023742988288  2016-05-13 16:15:54 +0000   
953   722974582966214656  2016-04-21 02:25:47 +0000   
990   716439118184652801  2016-04-03 01:36:11 +0000   
1016  713900603437621249  2016-03-27 01:29:02 +0000   
1041  710658690886586372  2016-03-18 02:46:49 +0000   
1061  709198395643068416  2016-03-14 02:04:08 +0000   
1136  704054845121142784  2016-02-28 21:25:30 +0000   
1218  697463031882764288  2016-02-10 16:51:59 +0000   
1410  684225744407494656  2016-01-05 04:11:44 +0000   
1411  684222868335505415  2016-01-05 04:00:18 +0000   
1552  677716515794329600  2015-12-18 05:06:23 +0000   
1615  675853064436391936  2015-12-13 01:41:41 +0000   

                                                                                  source  \
341   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
702   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
908   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
953   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
990   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1016  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1041  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1061  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1136  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1218  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1410  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1411  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1552  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1615  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                    text  \
341   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                  
702   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                       
908   Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                            
953   Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                     
990   This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                
1016  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                        
1041  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                         
1061  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1136  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                   
1218  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                         
1410  Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgiv

In [143]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2037 entries, 0 to 2116
Data columns (total 22 columns):
tweet_id                  2037 non-null int64
timestamp                 2037 non-null object
source                    2037 non-null object
text                      2037 non-null object
expanded_urls             2037 non-null object
name                      2037 non-null object
retweet_count             2035 non-null float64
favorite_count            2035 non-null float64
jpg_url                   1915 non-null object
img_num                   1915 non-null float64
p1                        1915 non-null object
p1_conf                   1915 non-null float64
p1_dog                    1915 non-null object
p2                        1915 non-null object
p2_conf                   1915 non-null float64
p2_dog                    1915 non-null object
p3                        1915 non-null object
p3_conf                   1915 non-null float64
p3_dog                    1915 non-null ob

#### Dog stage's type to categorical


##### Define

Convert *dog_stage* to categorical data type.

##### Code

In [144]:
# To category
df_twitter_archive_clean.dog_stage = df_twitter_archive_clean.dog_stage.astype('category')


##### Test

In [145]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2037 entries, 0 to 2116
Data columns (total 22 columns):
tweet_id                  2037 non-null int64
timestamp                 2037 non-null object
source                    2037 non-null object
text                      2037 non-null object
expanded_urls             2037 non-null object
name                      2037 non-null object
retweet_count             2035 non-null float64
favorite_count            2035 non-null float64
jpg_url                   1915 non-null object
img_num                   1915 non-null float64
p1                        1915 non-null object
p1_conf                   1915 non-null float64
p1_dog                    1915 non-null object
p2                        1915 non-null object
p2_conf                   1915 non-null float64
p2_dog                    1915 non-null object
p3                        1915 non-null object
p3_conf                   1915 non-null float64
p3_dog                    1915 non-null ob

#### Erroneous datatype (timestamp, tweet_id)

##### Define

Convert timestamp datetime data type and tweet_id to string data type.


##### Code

In [146]:
# To datetime
df_twitter_archive_clean.timestamp = pd.to_datetime(df_twitter_archive_clean.timestamp)

# To string 
df_twitter_archive_clean.tweet_id = df_twitter_archive_clean.tweet_id.astype(str)


In [147]:
df_twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2037 entries, 0 to 2116
Data columns (total 22 columns):
tweet_id                  2037 non-null object
timestamp                 2037 non-null datetime64[ns]
source                    2037 non-null object
text                      2037 non-null object
expanded_urls             2037 non-null object
name                      2037 non-null object
retweet_count             2035 non-null float64
favorite_count            2035 non-null float64
jpg_url                   1915 non-null object
img_num                   1915 non-null float64
p1                        1915 non-null object
p1_conf                   1915 non-null float64
p1_dog                    1915 non-null object
p2                        1915 non-null object
p2_conf                   1915 non-null float64
p2_dog                    1915 non-null object
p3                        1915 non-null object
p3_conf                   1915 non-null float64
p3_dog                    1915 no

In [148]:
df_twitter_archive_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

       name  retweet_count  favorite_count  \
0  Phineas   8315.0         38014.0          
1  Tilly     6142.0         32641.0          
2  Archie    4067.0         24565.0          
3  Darla     8450.0         41365.0          
4  Franklin  9157.0         39551.0          

                                           jpg_url  img_num  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg  1.0       
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg  1.0       
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg  1.0       
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg  1.0       
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg  2.0       

           ...           p1_dog                  p2   p2_conf p2_dog  \
0          ...            False  bagel               0.085851  False   
1          ...            True   Pekinese            0.090647  True    
2          ...            True   malamute            0.078253  True    
3          ...            False  Labrador_retriever  0.168086  True    
4          ...            True   English_springer    0.225770  True    

                            p3   p3_conf p3_dog  dog_stage  \
0  banana                       0.076110  False  None        
1  papillon                     0.068957  True   None        
2  kelpie                       0.031379  True   None        
3  spatula                      0.040836  False  N

#### Hard to understand the source from source column


##### Define

Make the source readable and clear in the *source* column by re-extracting the source with regex.

##### Code

In [149]:
df_twitter_archive_clean.source = df_twitter_archive_clean.source.apply(lambda x: re.findall(r'>(.*)<',x)[0])

In [150]:
df_twitter_archive_clean.source.value_counts()

Twitter for iPhone     1907
Vine - Make a Scene    90  
Twitter Web Client     30  
TweetDeck              10  
Name: source, dtype: int64

##### Test

In [151]:
df_twitter_archive_clean.sample(5)

tweet_id           timestamp              source  \
763   750117059602808832 2016-07-05 00:00:18  Twitter for iPhone   
75    876120275196170240 2017-06-17 16:52:05  Twitter for iPhone   
1207  698178924120031232 2016-02-12 16:16:41  Twitter for iPhone   
1270  692905862751522816 2016-01-29 03:03:25  Twitter for iPhone   
786   748575535303884801 2016-06-30 17:54:50  Twitter for iPhone   

                                                                                                                                                                      text  \
763   This is Calvin. He just loves America so much. 10/10 would roll around in flag with https://t.co/RXdzWaCQHm                                                            
75    Meet Venti, a seemingly caffeinated puppoccino. She was just informed the weekend would include walks, pats and scritches. 13/10 much excite https://t.co/ejExJFq3ek   
1207  This is Lily. She accidentally dropped all her Kohl's cash overboard. Day officially ruined. 10/10 hang in there pup https://t.co/BJbtCqGwZK                           
1270  This is Wally. He's being abducted by aliens. 10/10 poor pupper https://t.co/EiF659Bgjn                                                                                
786   This is one of the most reckless puppers I've ever seen. How she got a license in the first place is beyond me. 6/10 https://t.co/z5bAdtn9kd                           

                                                                                                                                                                                                                                                        expanded_urls  \
763   https://twitter.com/dog_rates/status/750117059602808832/photo/1,https://twitter.com/dog_rates/status/750117059602808832/photo/1,https://twitter.com/dog_rates/status/750117059602808832/photo/1,https://twitter.com/dog_rates/status/750117059602808832/photo/1   
75    https://twitter.com/dog_rates/status/876120275196170240/photo/1                                                                                                                                                                                                   
1207  https://twitter.com/dog_rates/status/698178924120031232/photo/1                                                                                                                                                                                                   
1270  https://twitter.com/dog_rates/status/692905862751522816/photo/1                                                                                                                                                                                                   
786   https://twitter.com/dog_rates/status/748575535303884801/photo/1                                                                                                                                                                                                   

        name  retweet_count  favorite_count  \
763   Calvin  1401.0         4567.0           
75    Venti   4611.0         27458.0          
1207  Lily    789.0          2930.0           
1270  Wally   934.0          2554.0           
786   None    2131.0         6374.0           

                                              jpg_url  img_num  \
763   https://pbs.twimg.com/media/Cmjzc-oWEAESFCm.jpg  2.0       
75    https://pbs.twimg.com/media/DCiavj_UwAAcXep.jpg  1.0       
1207  https://pbs.twimg.com/media/CbBuBhbWwAEGH29.jpg  1.0       
1270  https://pbs.twimg.com/media/CZ2yNKhWEAA_7cb.jpg  1.0       
786   https://pbs.twimg.com/media/CmN5ecNWMAE6pnf.jpg  1.0       

              ...           p1_dog             p2   p2_conf p2_dog  \
763           ...            True   Lhasa          0.175220  True    
75            ...            True   Saint_Bernard  0.346312  True    
1207          ...            True   malinois       0.207753  True    
1270       